README
NAME
    EV - perl interface to libev, a high performance full-featured event
    loop

SYNOPSIS
       use EV;
   
       # TIMERS
   
       my $w = EV::timer 2, 0, sub {
          warn "is called after 2s";
       };
   
       my $w = EV::timer 2, 2, sub {
          warn "is called roughly every 2s (repeat = 2)";
       };
   
       undef $w; # destroy event watcher again
   
       my $w = EV::periodic 0, 60, 0, sub {
          warn "is called every minute, on the minute, exactly";
       };
   
       # IO
   
       my $w = EV::io *STDIN, EV::READ, sub {
          my ($w, $revents) = @_; # all callbacks receive the watcher and event mask
          warn "stdin is readable, you entered: ", <STDIN>;
       };
   
       # SIGNALS
   
       my $w = EV::signal 'QUIT', sub {
          warn "sigquit received\n";
       };
   
       # CHILD/PID STATUS CHANGES
  
       my $w = EV::child 666, 0, sub {
          my ($w, $revents) = @_;
          my $status = $w->rstatus;
       };
  
       # STAT CHANGES
       my $w = EV::stat "/etc/passwd", 10, sub {
          my ($w, $revents) = @_;
          warn $w->path, " has changed somehow.\n";
       };
   
       # MAINLOOP
       EV::run;                # loop until EV::unloop is called or all watchers stop
       EV::run EV::RUN_ONCE;   # block until at least one event could be handled
       EV::run EV::RUN_NOWAIT; # try to handle same events, but do not block

BEFORE YOU START USING THIS MODULE
    If you only need timer, I/O, signal, child and idle watchers and not the
    advanced functionality of this module, consider using AnyEvent instead,
    specifically the simplified API described in AE.

    When used with EV as backend, the AE API is as fast as the native EV
    API, but your programs/modules will still run with many other event
    loops.

DESCRIPTION
    This module provides an interface to libev
    (<http://software.schmorp.de/pkg/libev.html>). While the documentation
    below is comprehensive, one might also consult the documentation of
    libev itself (<http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod> or
    perldoc EV::libev) for more subtle details on watcher semantics or some
    discussion on the available backends, or how to force a specific backend
    with "LIBEV_FLAGS", or just about in any case because it has much more
    detailed information.

    This module is very fast and scalable. It is actually so fast that you
    can use it through the AnyEvent module, stay portable to other event
    loops (if you don't rely on any watcher types not available through it)
    and still be faster than with any other event loop currently supported
    in Perl.

  PORTING FROM EV 3.X to 4.X
    EV version 4 introduces a number of incompatible changes summarised
    here. According to the depreciation strategy used by libev, there is a
    compatibility layer in place so programs should continue to run
    unchanged (the XS interface lacks this layer, so programs using that one
    need to be updated).

    This compatibility layer will be switched off in some future release.

    All changes relevant to Perl are renames of symbols, functions and
    methods:

      EV::loop          => EV::run
      EV::LOOP_NONBLOCK => EV::RUN_NOWAIT
      EV::LOOP_ONESHOT  => EV::RUN_ONCE

      EV::unloop        => EV::break
      EV::UNLOOP_CANCEL => EV::BREAK_CANCEL
      EV::UNLOOP_ONE    => EV::BREAK_ONE
      EV::UNLOOP_ALL    => EV::BREAK_ALL

      EV::TIMEOUT       => EV::TIMER

      EV::loop_count    => EV::iteration
      EV::loop_depth    => EV::depth
      EV::loop_verify   => EV::verify

    The loop object methods corresponding to the functions above have been
    similarly renamed.

  MODULE EXPORTS
    This module does not export any symbols.

EVENT LOOPS
    EV supports multiple event loops: There is a single "default event loop"
    that can handle everything including signals and child watchers, and any
    number of "dynamic event loops" that can use different backends (with
    various limitations), but no child and signal watchers.

    You do not have to do anything to create the default event loop: When
    the module is loaded a suitable backend is selected on the premise of
    selecting a working backend (which for example rules out kqueue on most
    BSDs). Modules should, unless they have "special needs" always use the
    default loop as this is fastest (perl-wise), best supported by other
    modules (e.g. AnyEvent or Coro) and most portable event loop.

    For specific programs you can create additional event loops dynamically.

    If you want to take advantage of kqueue (which often works properly for
    sockets only) even though the default loop doesn't enable it, you can
    *embed* a kqueue loop into the default loop: running the default loop
    will then also service the kqueue loop to some extent. See the example
    in the section about embed watchers for an example on how to achieve
    that.

    $loop = new EV::Loop [$flags]
        Create a new event loop as per the specified flags. Please refer to
        the "ev_loop_new ()" function description in the libev documentation
        (<http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#GLOBAL_FUNCTI
        ONS>, or locally-installed as EV::libev manpage) for more info.

        The loop will automatically be destroyed when it is no longer
        referenced by any watcher and the loop object goes out of scope.

        If you are not embedding the loop, then Using "EV::FLAG_FORKCHECK"
        is recommended, as only the default event loop is protected by this
        module. If you *are* embedding this loop in the default loop, this
        is not necessary, as "EV::embed" automatically does the right thing
        on fork.

    $loop->loop_fork
        Must be called after a fork in the child, before entering or
        continuing the event loop. An alternative is to use
        "EV::FLAG_FORKCHECK" which calls this function automatically, at
        some performance loss (refer to the libev documentation).

    $loop->verify
        Calls "ev_verify" to make internal consistency checks (for debugging
        libev) and abort the program if any data structures were found to be
        corrupted.

    $loop = EV::default_loop [$flags]
        Return the default loop (which is a singleton object). Since this
        module already creates the default loop with default flags,
        specifying flags here will not have any effect unless you destroy
        the default loop first, which isn't supported. So in short: don't do
        it, and if you break it, you get to keep the pieces.

BASIC INTERFACE
    $EV::DIED
        Must contain a reference to a function that is called when a
        callback throws an exception (with $@ containing the error). The
        default prints an informative message and continues.

        If this callback throws an exception it will be silently ignored.

    $flags = EV::supported_backends
    $flags = EV::recommended_backends
    $flags = EV::embeddable_backends
        Returns the set (see "EV::BACKEND_*" flags) of backends supported by
        this instance of EV, the set of recommended backends (supposed to be
        good) for this platform and the set of embeddable backends (see
        EMBED WATCHERS).

    EV::sleep $seconds
        Block the process for the given number of (fractional) seconds.

    $time = EV::time
        Returns the current time in (fractional) seconds since the epoch.

    $time = EV::now
    $time = $loop->now
        Returns the time the last event loop iteration has been started.
        This is the time that (relative) timers are based on, and referring
        to it is usually faster then calling EV::time.

    EV::now_update
    $loop->now_update
        Establishes the current time by querying the kernel, updating the
        time returned by "EV::now" in the progress. This is a costly
        operation and is usually done automatically within "EV::loop".

        This function is rarely useful, but when some event callback runs
        for a very long time without entering the event loop, updating
        libev's idea of the current time is a good idea.

    EV::suspend
    $loop->suspend
    EV::resume
    $loop->resume
        These two functions suspend and resume a loop, for use when the loop
        is not used for a while and timeouts should not be processed.

        A typical use case would be an interactive program such as a game:
        When the user presses "^Z" to suspend the game and resumes it an
        hour later it would be best to handle timeouts as if no time had
        actually passed while the program was suspended. This can be
        achieved by calling "suspend" in your "SIGTSTP" handler, sending
        yourself a "SIGSTOP" and calling "resume" directly afterwards to
        resume timer processing.

        Effectively, all "timer" watchers will be delayed by the time spend
        between "suspend" and "resume", and all "periodic" watchers will be
        rescheduled (that is, they will lose any events that would have
        occured while suspended).

        After calling "suspend" you must not call *any* function on the
        given loop other than "resume", and you must not call "resume"
        without a previous call to "suspend".

        Calling "suspend"/"resume" has the side effect of updating the event
        loop time (see "now_update").

    $backend = EV::backend
    $backend = $loop->backend
        Returns an integer describing the backend used by libev
        (EV::BACKEND_SELECT or EV::BACKEND_EPOLL).

    $active = EV::run [$flags]
    $active = $loop->run ([$flags])
        Begin checking for events and calling callbacks. It returns when a
        callback calls EV::unloop or the flasg are nonzero (in which case
        the return value is true) or when there are no active watchers which
        reference the loop (keepalive is true), in which case the return
        value will be false. The returnv alue can generally be interpreted
        as "if true, there is more work left to do".

        The $flags argument can be one of the following:

           0               as above
           EV::RUN_ONCE    block at most once (wait, but do not loop)
           EV::RUN_NOWAIT  do not block at all (fetch/handle events but do not wait)

    EV::break [$how]
    $loop->break ([$how])
        When called with no arguments or an argument of EV::BREAK_ONE, makes
        the innermost call to EV::loop return.

        When called with an argument of EV::BREAK_ALL, all calls to EV::loop
        will return as fast as possible.

        When called with an argument of EV::BREAK_CANCEL, any pending break
        will be cancelled.

    $count = EV::iteration
    $count = $loop->iteration
        Return the number of times the event loop has polled for new events.
        Sometimes useful as a generation counter.

    EV::once $fh_or_undef, $events, $timeout, $cb->($revents)
    $loop->once ($fh_or_undef, $events, $timeout, $cb->($revents))
        This function rolls together an I/O and a timer watcher for a single
        one-shot event without the need for managing a watcher object.

        If $fh_or_undef is a filehandle or file descriptor, then $events
        must be a bitset containing either "EV::READ", "EV::WRITE" or
        "EV::READ | EV::WRITE", indicating the type of I/O event you want to
        wait for. If you do not want to wait for some I/O event, specify
        "undef" for $fh_or_undef and 0 for $events).

        If timeout is "undef" or negative, then there will be no timeout.
        Otherwise a EV::timer with this value will be started.

        When an error occurs or either the timeout or I/O watcher triggers,
        then the callback will be called with the received event set (in
        general you can expect it to be a combination of "EV::ERROR",
        "EV::READ", "EV::WRITE" and "EV::TIMER").

        EV::once doesn't return anything: the watchers stay active till
        either of them triggers, then they will be stopped and freed, and
        the callback invoked.

    EV::feed_fd_event $fd, $revents
    $loop->feed_fd_event ($fd, $revents)
        Feed an event on a file descriptor into EV. EV will react to this
        call as if the readyness notifications specified by $revents (a
        combination of "EV::READ" and "EV::WRITE") happened on the file
        descriptor $fd.

    EV::feed_signal_event $signal
        Feed a signal event into the default loop. EV will react to this
        call as if the signal specified by $signal had occured.

    EV::feed_signal $signal
        Feed a signal event into EV - unlike "EV::feed_signal_event", this
        works regardless of which loop has registered the signal, and is
        mainly useful fro custom signal implementations.

    EV::set_io_collect_interval $time
    $loop->set_io_collect_interval ($time)
    EV::set_timeout_collect_interval $time
    $loop->set_timeout_collect_interval ($time)
        These advanced functions set the minimum block interval when polling
        for I/O events and the minimum wait interval for timer events. See
        the libev documentation at
        <http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#FUNCTIONS_CONT
        ROLLING_THE_EVENT_LOOP> (locally installed as EV::libev) for a more
        detailed discussion.

    $count = EV::pending_count
    $count = $loop->pending_count
        Returns the number of currently pending watchers.

    EV::invoke_pending
    $loop->invoke_pending
        Invoke all currently pending watchers.

WATCHER OBJECTS
    A watcher is an object that gets created to record your interest in some
    event. For instance, if you want to wait for STDIN to become readable,
    you would create an EV::io watcher for that:

       my $watcher = EV::io *STDIN, EV::READ, sub {
          my ($watcher, $revents) = @_;
          warn "yeah, STDIN should now be readable without blocking!\n"
       };

    All watchers can be active (waiting for events) or inactive (paused).
    Only active watchers will have their callbacks invoked. All callbacks
    will be called with at least two arguments: the watcher and a bitmask of
    received events.

    Each watcher type has its associated bit in revents, so you can use the
    same callback for multiple watchers. The event mask is named after the
    type, i.e. EV::child sets EV::CHILD, EV::prepare sets EV::PREPARE,
    EV::periodic sets EV::PERIODIC and so on, with the exception of I/O
    events (which can set both EV::READ and EV::WRITE bits).

    In the rare case where one wants to create a watcher but not start it at
    the same time, each constructor has a variant with a trailing "_ns" in
    its name, e.g. EV::io has a non-starting variant EV::io_ns and so on.

    Please note that a watcher will automatically be stopped when the
    watcher object is destroyed, so you *need* to keep the watcher objects
    returned by the constructors.

    Also, all methods changing some aspect of a watcher (->set, ->priority,
    ->fh and so on) automatically stop and start it again if it is active,
    which means pending events get lost.

  COMMON WATCHER METHODS
    This section lists methods common to all watchers.

    $w->start
        Starts a watcher if it isn't active already. Does nothing to an
        already active watcher. By default, all watchers start out in the
        active state (see the description of the "_ns" variants if you need
        stopped watchers).

    $w->stop
        Stop a watcher if it is active. Also clear any pending events
        (events that have been received but that didn't yet result in a
        callback invocation), regardless of whether the watcher was active
        or not.

    $bool = $w->is_active
        Returns true if the watcher is active, false otherwise.

    $current_data = $w->data
    $old_data = $w->data ($new_data)
        Queries a freely usable data scalar on the watcher and optionally
        changes it. This is a way to associate custom data with a watcher:

           my $w = EV::timer 60, 0, sub {
              warn $_[0]->data;
           };
           $w->data ("print me!");

    $current_cb = $w->cb
    $old_cb = $w->cb ($new_cb)
        Queries the callback on the watcher and optionally changes it. You
        can do this at any time without the watcher restarting.

    $current_priority = $w->priority
    $old_priority = $w->priority ($new_priority)
        Queries the priority on the watcher and optionally changes it.
        Pending watchers with higher priority will be invoked first. The
        valid range of priorities lies between EV::MAXPRI (default 2) and
        EV::MINPRI (default -2). If the priority is outside this range it
        will automatically be normalised to the nearest valid priority.

        The default priority of any newly-created watcher is 0.

        Note that the priority semantics have not yet been fleshed out and
        are subject to almost certain change.

    $w->invoke ($revents)
        Call the callback *now* with the given event mask.

    $w->feed_event ($revents)
        Feed some events on this watcher into EV. EV will react to this call
        as if the watcher had received the given $revents mask.

    $revents = $w->clear_pending
        If the watcher is pending, this function clears its pending status
        and returns its $revents bitset (as if its callback was invoked). If
        the watcher isn't pending it does nothing and returns 0.

    $previous_state = $w->keepalive ($bool)
        Normally, "EV::loop" will return when there are no active watchers
        (which is a "deadlock" because no progress can be made anymore).
        This is convenient because it allows you to start your watchers (and
        your jobs), call "EV::loop" once and when it returns you know that
        all your jobs are finished (or they forgot to register some watchers
        for their task :).

        Sometimes, however, this gets in your way, for example when the
        module that calls "EV::loop" (usually the main program) is not the
        same module as a long-living watcher (for example a DNS client
        module written by somebody else even). Then you might want any
        outstanding requests to be handled, but you would not want to keep
        "EV::loop" from returning just because you happen to have this
        long-running UDP port watcher.

        In this case you can clear the keepalive status, which means that
        even though your watcher is active, it won't keep "EV::loop" from
        returning.

        The initial value for keepalive is true (enabled), and you can
        change it any time.

        Example: Register an I/O watcher for some UDP socket but do not keep
        the event loop from running just because of that watcher.

           my $udp_socket = ...
           my $udp_watcher = EV::io $udp_socket, EV::READ, sub { ... };
           $udp_watcher->keepalive (0);

    $loop = $w->loop
        Return the loop that this watcher is attached to.

WATCHER TYPES
    Each of the following subsections describes a single watcher type.


Changes
Revision history for Perl extension EV

Changes marked with (libev) are changes in libev, and have more
documentation in the libev Changes file.

4.18 Sat Sep  6 20:37:23 CEST 2014
	- use slightly better weay to find includes, to support multiarch
          on newer perls.

4.17 Fri Apr 11 06:22:38 CEST 2014
	- perl5porters broke Async::Interrupt, BDB, EV, IO::AIO, OpenCL
          without warning by switching the meaning of USE_SOCKETS_AS_HANDLES
          in 5.18. What's so attractive about giving a shit about backwards
          compatibility - I will never understand.

4.16 Sat Mar  8 16:49:05 CET 2014
	- make sure EV::CHECK is available at all times.
        - avoid a spurious warning when perl -w overrides "no warnings".
        - use perl's signal implementation on WIN32.
        - ensure extern "C" function pointers are used for externally-visible
          pointers.
	- (libev) mark event pipe fd as cloexec after a fork (analyzed by Sami Farin).
        - (ecb) support m68k, m88k and sh (patch by Miod Vallat).
        - (libev) in the absence of autoconf, do not use the clock syscall
          on glibc >= 2.17 (avoids the syscall AND -lrt on systems
          doing clock_gettime in userspace).
        - (ecb) work around memory barriers and volatile apparently both being
          broken in visual studio 2008 and later (analysed and patch by Nicolas Noble).

4.15 Fri Mar  1 12:15:53 CET 2013
	- (libev) upgrade to 4.15 - too many changes to list.
	- EV::run now returns a boolean.
        - API version 5:1.
        - document that cleanup watchers are not available via perl.
        - cast I32 to int in error message printf.
        - remove dependencies on librt and libpthreads on GNU/Linux.

4.11 Sat Feb  4 19:56:26 CET 2012
	- (libev) implement memory fences for (obsolete) llvm-gcc.

4.10 Thu Jan 19 18:54:23 CET 2012
        - (libev) fix a race where the workaround against the epoll fork
          bugs caused signals to not be handled anymore.
        - (libev) correct backend_fudge for most backends, and implement
          a windows specific workaround to avoid looping because we call
          both select and Sleep, both with different time resolutions.
	- e_new wasn't declared static (causing very minor .so bloat).
        - replace more old api names by new ones.

4.03 Tue Jan 11 14:51:05 CET 2011
	- do not avoid the clock_gettime call on GNU/Linux anymore, as EV
          links against -lpthread anyways - as a result, EV might now take
          advantage of fast userspace clock_gettime implementations,
          but also links against -lrt.
        - (libev) lots of event port bug workarounds.
        - (libev) officially support files in I/O watchers.
        - (libev) new function ev_feed_signal.
        - fix documentation parts still refering to the 3.x API.

4.02 Thu Dec 30 08:27:41 CET 2010
        - the revents argument did not stringify correctly, as only the
          numeric value was updated, while the string value was kept from
          previous invocations.

4.01 Sun Dec  5 12:42:13 CET 2010
	- fully support EV_COMPAT3=0.
        - default_fork was stupidly defined as inline.
        - ask cpan to upgrade AnyEvent if < 5.29.
        - support EV_EXTRA_DEFS during configuration.
        - support -DEV_NO_LOOPS for snakker build.

4.00 Mon Oct 25 13:30:09 CEST 2010
	- many API changes, see the manual.
        - (libev) lots and lots of bugfixes, see the ev documentation.
	- fix a bug where inotify usage would parse the same event
          multiple times, causing various forms of breakage.
	- greatly reduce stack usage for inotify (8kb to <0.5kb).
        - expose ev_depth and ev_verify via the XS API.
        - implement ev_cleanup watchers.
        - (libev) ev_embed_stop did not correctly stop the watcher.
        - (libev) disable poll backend on AIX.
        - (libev) rename EV_TIMEOUT to EV_TIMER.
        - (libev) add section on accept() problems to the manpage.
        - (libev) no child watchers on win32.
        - make code more aliasing compliant, in case perl is ever
          translated to C.
        - document the EV::CHECK runtime unavailability.
        - ported to minix 3.1.7.

3.9  Thu Dec 31 07:59:59 CET 2009
	- disable t/07* under automatic testing.
        - increase t/09* timeouts as netbsd has *horribly* broken
          select/kevent that *usually* sleep >>0.6s instead of 0.3.
        - add constants EV::FLAG_NOSIGFD, EV::FLAG_NOINOTIFY and
          EV::BACKEND_ALL.
        - (libev) signalfd is no longer used by default.
	- (libev) backport inotify code to C89.
        - (libev) inotify file descriptors could leak into child processes.
        - (libev) ev_stat watchers could keep an errornous extra ref on the loop.
        - (libev) take advantage of inotify_init1, if available.
        - (libev) the signal handling pipe wasn't always initialised under windows.
        - changed minimum glibc requirement from glibc 2.9 to 2.7, for signalfd.
        - (libev) only replace ev_stat.prev when we detect an actual difference.

3.8  Sun Aug  9 15:30:10 CEST 2009
	- implement $loop->signal/signal_ns.
        - (libev) incompatible change: do not necessarily reset signal
          handler to SIG_DFL when a sighandler is stopped.
        - (libev) ev_default_destroy did not properly free or zero some
          members, potentially causing crashes and memory corruption on
          repated ev_default_destroy/ev_default_loop calls.
	- (libev) take advantage of signalfd on GNU/Linux systems.
        - (libev) document that the signal mask might be in an unspecified
          state when using libev's signal handling.
        - (libev) take advantage of some GNU/Linux calls to set
          cloexec/nonblock on fd creation, to avoid race conditions.
        - implement internal glue code to interface more efficiently
          with AnyEvent.

3.7  Fri Jul 17 16:49:16 CEST 2009
	- add EV::loop_verify and EV::loop_depth.

Makefile.PL
use 5.008002;

use strict qw(vars subs);
use Config;
use ExtUtils::MakeMaker;

sub have_inc($) {
   scalar grep -r "$_/$_[0]", $Config{usrinc}, split / /, $Config{incpth}
}

unless (-e "libev/ev_epoll.c") {
   print <<EOF;

***
*** ERROR: libev is missing or damaged. If you used a CVS check-out of EV,
*** you also have to check-out the "libev" module from the same CVS
*** repository into the EV dir (i.e. EV/libev from outside).
***

EOF
   exit 1;
}

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


Welcome to EV configuration. If you are in a hurry, just press return here
and hope for the best. The defaults should usually do.

EOF

if (prompt ("Skip further questions and use defaults (y/n)?", "y") =~ /[yY]/) {
   $ENV{PERL_MM_USE_DEFAULT} = 1;
}

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


POSIX optionally offers support for a monotonic clock source. EV
can take advantage of this clock source to detect time jumps more
reliably. Unfortunately, some systems are bound to be broken, so you can
disable this here: you can completely disable the detection and use of
the monotonic clock by answering 'n' here. Support for this clock type
will otherwise be autodetected at both compile- and runtime. (this setting
currently affects the use of nanosleep over select as well).

EOF

my $DEFINE .= " -DEV_USE_MONOTONIC=" . (0 + (prompt ("Enable optional support for CLOCK_MONOTONIC (y/n)?", "y") =~ /[yY]/));

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


POSIX optionally offers support for a (potentially) high-resolution
realtime clock interface. In a good implementation, using it is faster
than the normal method of using gettimeofday. Unfortunately, this option
is also bound to be broken on some systems, and current EV versions do not
actually call gettimeofday very often, so it defaults to no.

EOF

$DEFINE .= " -DEV_USE_REALTIME=" . (0 + (prompt ("Prefer clock_gettime (CLOCK_REALTIME) over gettimeofday (y/n)?", "n") =~ /[yY]/));

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


EV can use various backends with various portability issues. The select
backend is the most portable and makes for a good fallback, but it can be
limited to a low number of file descriptors and/or might not compile. If
you have problems with compiling ev_select.c, you might try to play around
with disabling it here, or forcing it to use the fd_set provided by your
OS, via the next question. I highly recommend keeping it in.

EOF

if (prompt ("Enable select backend (y/n)?", "y") =~ /[yY]/) {
   $DEFINE .= " -DEV_USE_SELECT=1";

   print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


The select backend can operate in two modes. One uses the system-provided
fd_set and is usually limited to 1024 file descriptors (64 on windows),
the other requires your header files to define NFDBITS and declare a
suitable fd_mask type. If you run into problems compiling ev_select.c, you
can try forcing the use of the system fd_set here.

EOF

   if (prompt ("Force use of system fd_set for select backend (y/n)?", "n") =~ /[yY]/) {
      $DEFINE .= " -DEV_SELECT_USE_FD_SET";
   }
} else {
   $DEFINE .= " -DEV_USE_SELECT=0";
}

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


The second very portable backend is poll(2). It does not exist on windows
and various versions of Mac OS X (and on the other versions it simply
doesn't work), but works basically everywhere else. It is recommended to use
the default here unless you run into compile problems in ev_poll.c.

EOF

$DEFINE .= " -DEV_USE_POLL=" . (0 + (prompt ("Enable poll backend (y/n)?", (have_inc "poll.h") ? "y" : "n") =~ /[yY]/));

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


Select and poll make it hard to write efficient servers, especially if the
number of active connections is much lower than the watched ones. GNU/Linux
systems have a more scalable method called "epoll", which EV can use. For
this to work, both your kernel and glibc have to support epoll, but if you
can compile it, the detection will be done at runtime, and EV will safely
fall back to using select when epoll isn't available. If unsure, accept
the default.

EOF

my $can_epoll = have_inc "sys/epoll.h";
$can_epoll = $ENV{EV_EPOLL} if exists $ENV{EV_EPOLL};
$DEFINE .= " -DEV_USE_EPOLL=" . (0 + (prompt ("Enable epoll backend (y/n)?", $can_epoll ? "y" : "n") =~ /[yY]/));

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


Similarly to the epoll backend above, EV can take advantage of kqueue on
many BSD systems. Support for kqueue will be detected at runtime, with a
safe fallback to other methods when it cannot be used.

Note that kqueue is broken on most operating systems, so by default it
won't be used on many platforms, but you can still create your own event
loop with kqueue backend if you ask specifically for it.

Here is what we know:

NetBSD:  partially working in at least 3.1 and later. Yeah! :)
FreeBSD: broken on at least 6.2-STABLE, spotty in later versions,
         sockets *likely* work, ptys definitely don't.
OpenBSD: reports indicate that it likely doesn't work
         (similar problems as on FreeBSD).
OS X:    completely, utterly broken on at least <= 10.6.

EOF

my $can_kqueue = have_inc "sys/event.h";
$can_kqueue = $ENV{EV_KQUEUE} if exists $ENV{EV_KQUEUE};
$DEFINE .= " -DEV_USE_KQUEUE=" . (0 + (prompt ("Enable kqueue backend (y/n)?", $can_kqueue ? "y" : "n") =~ /[yY]/));

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


Similarly to the kqueue backend above, EV can take advantage of the
solaris 10 event port interface. Support for event ports will be detected
at runtime, with a safe fallback to other methods when it cannot be used.

EOF

$DEFINE .= " -DEV_USE_PORT=" . (0 + (prompt ("Enable event port backend (y/n)?", (have_inc "sys/port.h") ? "y" : "n") =~ /[yY]/));

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


EV needs the functions pthread_atfork and clock_gettime. On most systems
you need some special libraries for this (such as -lrt and -lpthread). You
can specify additional libraries to provide these calls (and any other
required by EV) now, or accept the default.

On GNU/Linux systems, EV uses the LSB 3.1 __register_atfork function
to avoid the dependency on libpthread, and directly uses the clock_gettime
syscall to avoid a dependency on librt.

EOF

my $solaris_libs = $^O =~ /solaris/i ? "-lsocket -lnsl" : "";
my $librt = $^O =~ /linux/i ? "" : "-lpthread -lrt";
my $LIBS = exists $ENV{EV_LIBS} ? $ENV{EV_LIBS} : "$librt $solaris_libs";
$LIBS = prompt "Extra libraries for pthread_atfork and clock_gettime?", $LIBS;


print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


A backend of a different kind is the Linux inotify(7) interface, which can
be used to speed up (and reduce resource consumption) of stat watchers. If
you have the include file and libc support for it, it is usually a good
idea to enable it, as kernel availability is detected at runtime.

EOF

my $can_inotify = have_inc "sys/inotify.h";
$can_inotify = $ENV{EV_INOTIFY} if exists $ENV{EV_INOTIFY};
$DEFINE .= " -DEV_USE_INOTIFY=" . (0 + (prompt ("Enable inotify support (y/n)?", $can_inotify ? "y" : "n") =~ /[yY]/));

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


Another useful bit of functionality is the Linux eventfd, which is useful
for faster signal handling (don't care) and intra-thread communications
(more relevant). Kernel support for this will be probed at runtime, but
your libc must contain the necessary wrapper. Glibc 2.7 and later should
have this wrapper.

EOF

my $can_eventfd = have_inc "sys/eventfd.h";
$can_eventfd = $ENV{EV_EVENTFD} if exists $ENV{EV_EVENTFD};
$DEFINE .= " -DEV_USE_EVENTFD=" . (0 + (prompt ("Enable linux eventfd support (y/n)?", $can_eventfd ? "y" : "n") =~ /[yY]/));

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


Another sometimes useful bit of functionality is the Linux signalfd, which
is useful for faster signal handling (don't care). Kernel support for
this will be probed at runtime, but your libc must contain the necessary
wrapper. Glibc 2.7 and later should have this wrapper.

EOF

my $can_signalfd = have_inc "sys/signalfd.h";
$can_signalfd = $ENV{EV_SIGNALFD} if exists $ENV{EV_SIGNALFD};
$DEFINE .= " -DEV_USE_SIGNALFD=" . (0 + (prompt ("Enable linux signalfd support (y/n)?", $can_signalfd ? "y" : "n") =~ /[yY]/));

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


Very rarely, people want to tweak EV even more, e.g. to exclude
or include certain watcher types or backends. This can be done by adding
extra -D options here, or via the EV_EXTRA_DEFS environment variable.

For example, if you run into compile problems because of missing memory
fences (or you just want extra performance), you can tell EV to not support
smp and threads via -DEV_NO_THREADS.

Normal persons just press enter.

EOF

$DEFINE .= " " . prompt "Any extra -D options?", "$ENV{EV_EXTRA_DEFS}";

print <<EOF;

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***


EOF

my @anyevent = eval { require AnyEvent; $AnyEvent::VERSION < 5.29 } ? (AnyEvent => 5.29) : ();

WriteMakefile(
    dist	=> {
	            PREOP	=> 'pod2text EV.pm | tee README >$(DISTVNAME)/README; chmod -R u=rwX,go=rX . ;',
	            COMPRESS	=> 'gzip -9v',
	            SUFFIX	=> '.gz',
	           },
    depend => {
       "EV.c" => "EV/EVAPI.h "
               . "libev/ev.c libev/ev.h libev/ev_epoll.c libev/ev_select.c libev/ev_kqueue.c libev/ev_poll.c "
               . "libev/ev_vars.h libev/ev_wrap.h",
    },
    INC       => "-Ilibev",
    DEFINE    => "$DEFINE",
    NAME => "EV",
    LIBS => [$LIBS],
    PREREQ_PM => {
       @anyevent,
       "common::sense" => 0,
    },
    VERSION_FROM => "EV.pm",
    PM           => {
		    'EV.pm'		=> '$(INST_LIB)/EV.pm',
		    'EV/EVAPI.h'	=> '$(INST_LIB)/EV/EVAPI.h',
		    'EV/MakeMaker.pm'	=> '$(INST_LIB)/EV/MakeMaker.pm',
		    'libev/ev.h'	=> '$(INST_LIB)/EV/ev.h',
		    'libev/ev.pod'	=> '$(INST_LIB)/EV/libev.pod',
                 },
    MAN3PODS => {
                    'EV.pm'           => '$(INST_MAN3DIR)/EV.$(MAN3EXT)',
                    'EV/MakeMaker.pm' => '$(INST_MAN3DIR)/EV::MakeMaker.$(MAN3EXT)',
                    'libev/ev.pod'    => '$(INST_MAN3DIR)/EV::libev.$(MAN3EXT)',
                },
);


